### DISCUSSION TITLE
能否透過Proxy連線呢?

### DISCUSSION THREAD

**Discussion by lovehearn** [hgalytoby/MyselfAnimeDownloader#27] [Q&A] [answered]
請問如何透過Proxy連線?
我是使用windows bainary版的 在某些環境無法直接連接到myself的網站 
一開始的程式執行就會被擋下來了，要如何設定才能讓程式透由proxy連線呢?

**Comment by hgalytoby** [comment, marked as answer]
> 請問如何透過Proxy連線? 我是使用windows bainary版的 在某些環境無法直接連接到myself的網站 一開始的程式執行就會被擋下來了，要如何設定才能讓程式透由proxy連線呢?

你的意思是使用原始碼跑嗎？

**Reply by lovehearn** [reply]
我是如何通過代理連接線的？手動設定prxoy位址
我是如何直接使用windows 連接線的 ？ 一般的網路線阿 
我想問的是 能否透過執行參數的設定 這類的方式設定 讓MyselfAnime.exe 能夠透由指定的proxy連線
如果是透由原始檔 是可以透由設定 https_proxy的環境變數 即可達到效果

**Reply by hgalytoby** [reply]
> 我是如何通過代理連接線的？手動設定prxoy位址 我是如何直接使用windows 連接線的 ？ 一般的網路線阿 我想問的是 能否透過執行參數的設定 這類的方式設定 讓MyselfAnime.exe 能夠透由指定的proxy連線 如果是透由原始檔 是可以透由設定 https_proxy的環境變數 即可達到效果


不好意思我在公司看到訊息的時候，訊息怪怪的😅，我有修改訊息了。

我一開始不懂 binary 是什麼意思，現在懂了。

我的程式沒有 proxy 功能，我覺得應該是可以做到這個功能的，但我不知道怎麼測試 proxy。

**Reply by lovehearn** [reply]
> > 我是如何通過代理連接線的？手動設定prxoy位址 我是如何直接使用windows 連接線的 ？ 一般的網路線阿 我想問的是 能否透過執行參數的設定 這類的方式設定 讓MyselfAnime.exe 能夠透由指定的proxy連線 如果是透由原始檔 是可以透由設定 https_proxy的環境變數 即可達到效果
> 
> 不好意思我在公司看到訊息的時候，訊息怪怪的😅，我有修改訊息了。
> 
> 我一開始不懂 binary 是什麼意思，現在懂了。
> 
> 我的程式沒有 proxy 功能，我覺得應該是可以做到這個功能的，但我不知道怎麼測試 proxy。
1.可以透由無界瀏覽測試   https://ultrasurf.us/  windows上直接執行就會幫忙做好簡易設定了
2.利用Charles-proxy 這類工具 建立本地proxy 也可以測試

**Reply by hgalytoby** [reply]
我稍微研究了 ultrasurf 與 Charles-proxy。

 不知道有沒有理解錯誤。

### ultrasurf
VPN，連線後會變成其他 IP。

### Charles-proxy
攔截封包用途。

如果我沒理解錯誤，你使用 ultrasurf 後換了 IP，能看到 MyselfAnime 網站後再開啟程式不就好了嗎？

還是 ultrasurf 有個功能會提供 ip 與 port 後讓你使用？然後你自定義 proxy 後，向 MyslefAnime 發請求就會變成自定義的 proxy 去請求的？

所以你的原本網路請求還是自己網路，只有 MyselfAnime 是 Proxy 去請求的？

**Reply by lovehearn** [reply]
比較好的測試方式還是透過 Charles 自定proxy的方式比較好 因為可以看到程式是否有透過proxy流動 並且可以看到有哪些連線
他會自建一個proxy server 讓App可以透由這個proxy 連線 並追蹤封包
 
Q. 還是 ultrasurf 有個功能會提供 ip 與 port 後讓你使用？
YES  預設是 http://127.0.0.1:9666
Q.然後你自定義 proxy 後，向 MyslefAnime 發請求就會變成自定義的 proxy 去請求的？
預設系統只會幫忙設定系統proxy(也可以取消此行為) 但是除了瀏覽器外 一般應用程式通常並不會去自動透由系統設定的proxy連線 需要自己在程式中設定proxy 才可以
Q. 所以你的原本網路請求還是自己網路，只有 MyselfAnime 是 Proxy 去請求的？
基本上是這樣沒錯 是只有設定的app才會透由自定proxy連線
我原本網路環境 會擋myself-bbs 所以才需要透由proxy 或是VPN連線
 但是目前我用的免費proxy感覺通常都很不穩 速度也很慢 只是堪用而已

**Reply by hgalytoby** [reply]
我是不是只要準備一個攔截封包程式與一個 Proxy 程式，然後測試我發請求是不是改成 Proxy IP 就可以了？

我看不太懂 Charles 用法，是不是與 Charles-proxy 不一樣？

我想到一個測試方法。

我回家下載 ultrasurf ，然後用架伺服器，接著用 ultrasurf 提供的 IP PORT 去發請求給伺服器，看 IP 是不是變成 ultrasurf 就好了。

你覺得這樣測試如何？

**Reply by hgalytoby** [reply]
我只要一打開 usf.exe 程式，我整台電腦請求都變成 proxy 了，

![image](https://user-images.githubusercontent.com/50397689/188661380-f12eb4ac-acc7-440f-9fc2-2008cb147173.png)

不需要特別設定 proxy 吧?

不過我看 requests 會一直報錯就是了，下載會很慢。

**Reply by lovehearn** [reply]
我只要一打開 usf.exe 程式，我整台電腦請求都變成 proxy 了>>> 連MyselfAnime也是嗎?你MyselfAnime是用哪一版? 是win10嗎? 
requests 會一直報錯就是了，下載會很慢。>>的確是這樣 這之後可透過比較高級的VPN改善

**Reply by hgalytoby** [reply]
> 連MyselfAnime也是嗎?你MyselfAnime是用哪一版? 是win10嗎? 

![image](https://user-images.githubusercontent.com/50397689/188672844-87d2f0d6-bc39-4c08-bde8-389d1b28d91a.png)

MyselfAnime 就目前我做的最新版。

> 的確是這樣 這之後可透過比較高級的VPN改善

我只是想確認這種程式是不是只要一打開整個電腦請求都變了，還是只有我是這樣?

我剛剛嘗試了一下，我只要在每個 requests 都加個 try 就可以解決程式打不開的問題。

因為首頁請求我都沒加 try，所以只要訪問錯了就會報錯關閉程式。

但是下載會很慢而且方法很暴力就是了，這是我剛剛測了一個小時後的解法= =

**Reply by lovehearn** [reply]
你是用甚麼指令編譯成exe檔 我用最新原始檔試看看

**Reply by hgalytoby** [reply]
你要用原始碼跑直接 run `main.py` 就好了。

如果你要打包成 exe，`pyinstaller -F main.py`。

其餘指令你可以在 `指令.txt` 找到。

**Reply by lovehearn** [reply]
我剛試過了用自行編譯的程式的確是可以正常執行 並且抓到系統環境proxy設定  看來是之前release的版本跟現行的原始碼有些地方可能處理方式不一樣?

**Reply by hgalytoby** [reply]
你裝的套件版本都跟我的 `requirements.txt` 一致嗎？

**Reply by hgalytoby** [reply]
我在測試的時候 proxy 的時候，使用的 requests 版本比較舊，此專案用的比較新，導致我一直報錯程式關閉。

後來我降低 requests 版本後，就會偶爾發首頁請求不會錯誤的問題，但還是有機會打開時錯誤。

因為我的程式只有在下載影片的 requests 有包 try，其餘的 requests MyselfAnime 都沒有包，所以只要基本的請求錯了，我的程式會直接關閉。

**Reply by lovehearn** [reply]
> 我剛剛嘗試了一下，我只要在每個 requests 都加個 try 就可以解決程式打不開的問題。
> 
> 因為首頁請求我都沒加 try，所以只要訪問錯了就會報錯關閉程式。
訪問錯了就會報錯關閉程式 >> 一般程式 可能還是要有預設會有錯誤的情況會比較理想
另外有錯誤留下log 也比較好追蹤問題
總之問題不大 看最後是否想要如何做的更完善而已 
我目前也是先用原始檔執行  之前我也有發生過編碼錯誤的問題 目前還無法重現問題 無法回報狀況
上次我是直接把log history undone那些都砍掉 就可以正常執行了

**Comment by hgalytoby** [comment]
我最初開發的時候沒有想到會有看不到網站的問題。

後來因為有人反應使用 VPN 後才能打開程式，所以我臨時加了開啟程式時會去 request 官網。

沒有想過有人打開程式時檢查官網 request 成功後，後續的 request  官網資料會失敗。

我在測試 ultrasurf 的時候我用瀏覽器開啟官網看什麼都沒有問題，結果我用 Python requests 時會有問題。

最奇怪的是我新開一個 project 環境是乾淨的，我連續發了很多官網各種網頁的 request 都是沒有問題的。

結果我把程式碼放到 MyselfAnimeDownloader 專案時，發 request  就會報錯，我完全問號問號。

所以我最後才說我花了一個小時的解法是在所有 request 前面加 try，try 前面加 While。

**Reply by lovehearn** [reply]
> **我在測試 ultrasurf 的時候我用瀏覽器開啟官網看什麼都沒有問題，結果我用 Python requests 時會有問題。**

這就是我說的瀏覽器本身會自動抓取windows 系統的proxy設定 其他一般程式需要另外設定 
python 使用原始法執行時 是可以設定https_proxy環境變數 或是手動抓取系統proxy後 設定proxy連線
但是如果是網路環境有另外的問題導致連線失敗，這種狀況的確是可以考慮重新連線 
或許可以重新封裝request 令其可以自動重新連線 
並且可在自定 重試次數上限 timeout時間  是否要跳出視窗提醒是否繼續嘗試( 我們看一般的下載軟體都有類似的設計)
這樣就不用在每個request都使用 try例外處理 
提供一點建議參考看看

**Reply by hgalytoby** [reply]
這專案是我菜鳥沒工作的時候寫的。

如果要做優化，我想直接打掉重寫。

但是我已經朝向 Web 技術了，目前不打算花時間重寫 Qt，現在我只有維護 Bug 或加功能。

如果我真的要花時間重寫這個專案，我會順便整合 Anime1.me 的網站。

我今年初有嘗試爬 Anime1.me 的動漫，已經了解影片 Api 了，但是當時每過幾星期 Api 就變了，後來就不研究了 。

**Reply by lovehearn** [reply]
所以說**是否要提供Proxy設定** 和 **連線狀況的例外處裡**是兩件不太一樣的需求
如果有proxy設定參數 則可以根據不同需求 連線至不同的proxy 不用跟系統綁在一起 
甚至有些網站是一定要透由proxy連線的  相信你如果去找就可以找到很多方法讓 request透由proxy連線的設定的
https://www.scrapingbee.com/blog/python-requests-proxy/

還是很感謝你花時間設計這個軟體免費給大家使用喔  ^^

**Reply by lovehearn** [reply]
恩恩 一切就再看因緣了 
我以前也都只是開發底層系統而已  這類需要UI的App 也是要重新學習 有點懶惰 目前能用就好了
 Anime1.me 上的確也是有很多好東西啊  ^^ 
我之前是用 https://github.com/sakuraakira/AniGamerDownloader 這個工具下載 他的設計也是蠻有趣的
你看它就有proxy設定 另外有些是給cookie模擬瀏覽器行為 代替帳號密碼登入 安全性會更高一點 在有些無法重複登入的系統就很好用

**Comment by hgalytoby** [comment]
我知道怎麼在 requests 使用 proxy，只是 ultrasurf 一打開就把我只個系統 proxy 都蓋過去了，我根本沒辦法測試。

巴哈姆特我也爬過了，只是因為需要使用者手動填入 cookie 就沒做了，況且也有現成的 https://github.com/miyouzi/aniGamerPlus。
